### DISCUSSION TITLE
problems using -bilateral-blur

### DISCUSSION THREAD

**Discussion by DanielDrs** [ImageMagick/ImageMagick#5786] [Development]
Hallo,

I observed problems using -bilateral-blur.  

I executed

convert board.tif -bilateral-blur 50x50+10+10 -type truecolor test17.tif

using ImageMagick 7.1.0-17 Q16-HDRI x64.   In the same way, I created  
test52.tif using ImageMagick 7.1.0-52 Q16-HDRI x64.

The images test17.tif and test52.tif differ.  By executing the above commands 
several times using the same version of ImageMagick, I got several different 
images test??.tif.

I executed

convert board.tif -bilateral-blur 20x20+20+20 -type truecolor test52a.tif

using ImageMagick 7.1.0-52 Q16-HDRI x64.     Just consider the arrow in 
test52a-marked.tif.    The values decrease from 6416 to 6338 along the arrow. 
However, the values should increase along the arrow.

https://www.dropbox.com/sh/yzpne92zvisgcnu/AAApDXpU3j7VyvYjCx_RaIPTa?dl=0

Best regards, Daniel

**Comment by fmw42** [comment]
post your input image so we can test with it

**Reply by DanielDrs** [reply]
The input image is available under the dropbox-link.  I was not able to post tif-images.

**Comment by fmw42** [comment]
For IM 7, I suggest you use magick not convert.

I ran the following 10 times on ImageMagick 7.1.0-52 Q16-HDRI Mac OSX Monterey and got the exact same result each time as indicated by the 0 (0) results of compare. 


```
for ((i=0; i<10; i++)); do
echo $i
magick board.tif -bilateral-blur 20x20+20+20 -type truecolor test52_$i.tif
done


for ((i=1; i<10; i++)); do
j=$((i-1))
magick compare -metric rmse test52_$i.tif test52_$j.tif null:
echo ""
done

0 (0)
0 (0)
0 (0)
0 (0)
0 (0)
0 (0)
0 (0)
0 (0)
0 (0)

```

**Comment by DanielDrs** [comment]
Hallo,

the different outputs occurred when I use the paramters  50x50+10+10.     You should wait between the runs, e.g. make one run, 
do some other fancy stuff with you computer, and then, make another run.    I seems to me that some variable is not initialized.

Could you provide a link where I can learn something about the differences between convert and magick?

Best regards, Daniel

**Reply by fmw42** [reply]
If you have a symbolic link between convert and magick, then both are fine. But if not, then convert will use IM 6 syntax and will not get the IM 7 improvements. So best to use IM 7 to get all the new features.  See https://imagemagick.org/script/porting.php#cli

**Comment by fmw42** [comment]
I still do not see any differences even with

```
for ((i=0; i<10; i++)); do
echo $i
magick board.tif -bilateral-blur 50x50+10+10 -type truecolor test52_$i.tif
sleep 10
magick lena.jpg x.png
done


for ((i=1; i<10; i++)); do
j=$((i-1))
magick compare -metric rmse test52_$i.tif test52_$j.tif null:
echo ""
done

0 (0)
0 (0)
0 (0)
0 (0)
0 (0)
0 (0)
0 (0)
0 (0)
0 (0)

```

**Comment by snibgo** [comment]
As fmw42 says, I suggest using "magick", not "convert". But I doubt that that will cure your problem.

Please show the text output from "magick -version". Also, what is your platform?

There is nothing in "-bilateral-blur" that should give varying results. Sometimes OpenCL can do this, so if you have that, try disabling it.

Another possibility is a bad build of IM. Did you build it yourself? If so, how?

**Comment by DanielDrs** [comment]
magick -version
Version: ImageMagick 7.1.0-52 Q16-HDRI x64 04ee6ce:20221106 https://imagemagick.
org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenCL
Delegates (built-in): bzlib cairo freetype gslib heic jng jp2 jpeg jxl lcms lqr
lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193331630)

I downloaded   ImageMagick-7.1.0-portable-Q16-HDRI-x64.zip from  https://imagemagick.org/script/download.php

I have Windows 10 and run magick, convert, ...  from Cygwin.

**Comment by DanielDrs** [comment]
I executed

magick board.tif -bilateral-blur 50x50+10+10 -type truecolor test52m.tif

twice.   I got two different results, test52m.tif and test52n.tif which are avaliable 
under the above dropbox link.

Daniel

**Reply by fmw42** [reply]
As user snibgo said, try disabling OpenCL and see if that helps.

**Reply by DanielDrs** [reply]
How can I disable OpenCL  (OpenCV?)  in the portable Win10 version without compiling IM?

**Reply by snibgo** [reply]
IM doesn't use OpenCV. It can be built to use OpenCL, and has been in this case.

**Comment by snibgo** [comment]
To prevent IM from using OpenCL, this Windows command, which will have effect only for the session:
```
set MAGICK_OCL_DEVICE=off
```
... then try the commands that give problems. If the problems stop, then make the change permanent (Using Windows Control Panel, or some other mechanism).

**Comment by DanielDrs** [comment]
I tried, but I got another different strange output image.

D:\t\gmic-test>set MAGICK_OCL_DEVICE=off

D:\t\gmic-test>c:\Users\Danie\Downloads\ImageMagick-7.1.0-portable-Q16-HDRI-x64\magick.exe -version
Version: ImageMagick 7.1.0-52 Q16-HDRI x64 04ee6ce:20221106 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenCL
Delegates (built-in): bzlib cairo freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193331630)

D:\t\gmic-test>c:\Users\Danie\Downloads\ImageMagick-7.1.0-portable-Q16-HDRI-x64\magick.exe   board.tif -bilateral-blur 50x50+10+10 -type truecolor test52q.tif

**Comment by DanielDrs** [comment]
I compiled IM under cygwin by myself:

Version: ImageMagick 7.1.0-53 beta Q16-HDRI x86_64 97d1574ba:20221120 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5)
Delegates (built-in): jbig jng jpeg lzma png tiff webp zlib
Compiler: gcc (11.3)

The same runs are much faster,  the CPU consumption is higher  (accoriding to the fan), 
and the problem disappeared.
